tools/system_libs.py

```diff
@@ -1437,23 +1437,27 @@ def fetch_project(name, url, subdir, is_tarbz2=False, sha512hash=None):
     # clears the build, so that it is rebuilt from that source.
     local_ports = os.environ.get('EMCC_LOCAL_PORTS')
     if local_ports:
-      logger.warning('using local ports: %s' % local_ports)
-      local_ports = [pair.split('=', 1) for pair in local_ports.split(',')]
-      for local in local_ports:
-        if name == local[0]:
-          path = local[1]
-          if name not in ports.ports_by_name:
-            shared.exit_with_error('%s is not a known port' % name)
-          port = ports.ports_by_name[name]
-          if not hasattr(port, 'SUBDIR'):
-            logger.error('port %s lacks .SUBDIR attribute, which we need in order to override it locally, please update it' % name)
-            sys.exit(1)
-          subdir = port.SUBDIR
-          logger.warning('grabbing local port: ' + name + ' from ' + path + ' to ' + fullname + ' (subdir: ' + subdir + ')')
-          shared.try_delete(fullname)
-          shutil.copytree(path, os.path.join(fullname, subdir))
-          Ports.clear_project_build(name)
-          return
+      shared.Cache.acquire_cache_lock()
+      try:
+        logger.warning('using local ports: %s' % local_ports)
+        local_ports = [pair.split('=', 1) for pair in local_ports.split(',')]
+        for local in local_ports:
+          if name == local[0]:
+            path = local[1]
+            if name not in ports.ports_by_name:
+              shared.exit_with_error('%s is not a known port' % name)
+            port = ports.ports_by_name[name]
+            if not hasattr(port, 'SUBDIR'):
+              logger.error('port %s lacks .SUBDIR attribute, which we need in order to override it locally, please update it' % name)
+              sys.exit(1)
+            subdir = port.SUBDIR
+            logger.warning('grabbing local port: ' + name + ' from ' + path + ' to ' + fullname + ' (subdir: ' + subdir + ')')
+            shared.try_delete(fullname)
+            shutil.copytree(path, os.path.join(fullname, subdir))
+            Ports.clear_project_build(name)
```

Comment:
Presumably its these three lines that require the lock?

Reply:
The SDL2 port deletes depending ports prior (re)building, so I'm afraid that locking only those lines will lead to build failures like SDL2 (process1) > SDL2_image (p1) > SDL2 (process2) > SDL2_ttf (p1) > missing SDL2.h.
That's basically why I didn't try to be specific about the lock.
I cannot confirm right now due to #9402.

Reply:
This seems reasonable to me. Basically, any local-ports work should be locked.